### ISSUE TITLE
Reduce indexing time for syncing dataset use-case

### ISSUE THREAD

**Issue by gmourier** [meilisearch/product#520]
## Description

Users may regularly re-index at a given time (e.g each day/ each week) a dataset to ensure that an index is fully synchronized with the last seen state of their primary data source/data store.

Some reasons for that:

- It's difficult to manage each document change over time.
- If a solution is implemented to detect changes, hooks may be missing in the code and the index will get out of sync.

Meilisearch does not detect that the exact same version is already present and will go through the process as if the document had never existed and thus imply an indexing time for it when there is no reason to do so.

## Solution

It is possible to implement a mechanism to detect that a document representation already exists to short-circuit the indexing process.

**Comment by gmourier** [comment]
Implemented by https://github.com/meilisearch/milli/pull/579

Note: If the document fields are not returned in the same order as the previously indexed document. The short circuit will not be triggered, and the given document will be fully re-indexed.

**Comment by Prashank25** [comment]
I am wondering if you can sort the document by keys and then do the comparison? Order of fields being different causing a reindexing seems like something that will confuse people.

**Comment by irevoire** [comment]
Hello @Prashank25, we decided not to do anything complicated in this pipeline because that will affect ALL indexing operations and thus slow them down.
But FYI, the top level fields of the documents are already sorted, which means;

```JSON
{
  "id": 1,
  "content": "toto"
}
==
{
  "content": "toto",
  "id": 1
}
```

But,
```JSON
{
  "embed": {
    "id": 1,
    "content": "toto"
  }
}
!=
{
  "embed": {
    "content": "toto",
    "id": 1
  }
}
```

I hope that clarifies the situation a little bit :+1:

**Comment by gmourier** [comment]
Closed with `v0.29` release.
